File: reverse-linked-list/golang/reverseLinkedList.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var stack []ListNode
	nextNode := head

	for {
		if nextNode == nil {
			break
		}
		stack = append(stack, *nextNode)
		nextNode = nextNode.Next
	}

	var newHead *ListNode
	var currentNode *ListNode

	for len(stack) > 0 {
		var popped ListNode
		popped, stack = stack[len(stack)-1], stack[:len(stack)-1]

		if newHead == nil {
			newHead = &ListNode{popped.Val, nil}
			currentNode = newHead
			continue
		}

		currentNode.Next = &popped
		currentNode = currentNode.Next
		currentNode.Next = nil
	}

	return newHead
}

func main() {
	fmt.Println("Placeholder")
}

File: reverse-linked-list/typescript/reverseLinkedList.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

export type ListNode = {
  val: number,
  next: ListNode | null
}

export default function reverseList(head: ListNode | null): ListNode | null {
  let stack: ListNode[] = [];
  let nextNode: ListNode | null = head;

  while (true) {
    if(nextNode === null) { break }
    stack.push(nextNode);
    nextNode = nextNode.next;
  }

  let newHead: ListNode | null = null;
  let currentNode: ListNode | null = null;

  while(stack.length > 0) {
    const popped = stack.pop();
    if(newHead === null) {
      newHead = popped!;
      currentNode = newHead
      continue
    }

		currentNode!.next = popped!
		currentNode = currentNode!.next
		currentNode!.next = null
  }

  return newHead;
};
